### DISCUSSION TITLE
[Bug]: How to extend properties of FabricObject (fabric base object)

### DISCUSSION THREAD

**Discussion by shaktippatel** [fabricjs/fabric.js#8846] [Q&A]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have read and followed the [Issue Tracker Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md#%EF%B8%8F-issue-tracker)
- [X] I have searched and referenced existing issues and discussions
- [X] I am filing a **BUG** report.
- [X] I have managed to reproduce the bug **after** upgrading to the **latest** version
- [X] I have created an accurate and minimal reproduction

### Version

6.0.0-beta3

### Node Version (if applicable)

19.8.1

### Link To Reproduction

Next template >= v6

### Steps To Reproduce

In previous version we are extending base object properties with below code but not found extending FabricObject (fabric base object) in new version **6.0.0-beta3**

**OLD Code**
```
fabric.Object.prototype.toObject = (function (toObject) {
    return function (propertiesToInclude) {
      return toObject.call(this, ['uniqueId'].concat(propertiesToInclude));
    };
})(fabric.Object.prototype.toObject);
```

### Expected Behavior

Fabric base object **propertiesToInclude** should be extendable.

### Actual Behavior

Not able to extend Fabric base object **propertiesToInclude**.

**Comment by ShaMan123** [comment]
you didn't provide a reproduction

**Reply by shaktippatel** [reply]
`
import { Object as FabricObject} from "fabric";

FabricObject.prototype.toObject = ((toObject) => {
  return (propertiesToInclude:string[]) => {
    return toObject.call(this, ['uniqueId'].concat(propertiesToInclude));
  };
})(FabricObject.prototype.toObject);
`
**I have tried above code with new version but toObject method is protected so not able to override.**

**Reply by ShaMan123** [reply]
This should work as before

**Reply by shaktippatel** [reply]
Not working due to protected method

**Reply by shaktippatel** [reply]
Can you plz show example how to extend object with toObject method

**Reply by ShaMan123** [reply]
I don't see a problem
When saying reproduction we mean this:
https://codesandbox.io/p/sandbox/fabric-nextjs-sandbox-forked-z7l2nx?file=%2Fpages%2Findex.tsx&selection=%5B%7B%22endColumn%22%3A32%2C%22endLineNumber%22%3A25%2C%22startColumn%22%3A32%2C%22startLineNumber%22%3A25%7D%5D

**Reply by shaktippatel** [reply]
In previous version we are creating new object with below code but not able to understand with fabricjs v6

**OLD Code**

```
fabric.Square = fabric.util.createClass(fabric.Rect, {
  type: 'square',
  customBorder: false,
  initialize : function(options) {
      this.callSuper('initialize', options);
  },
  toObject: function(propertiesToInclude) {
    return this.callSuper('toObject', ['customBorder'].concat(propertiesToInclude));
  }
});
```


**New Code**
```
import * as fabric from "fabric";
export class Square extends fabric.Rect {
  customBorder: boolean = false;
  constructor(options: Record<string, unknown>){
    super(options);
  }
}
Object.assign(Square, {type: "square"});

Square.prototype.toObject = (function (toObject) {
  return function (this: Square, propertiesToInclude: any[] = []) {
    return toObject.call(this, [["customBorder"], ...propertiesToInclude] );
  };
})(fabric.Rect.prototype.toObject);

fabric.classRegistry.setClass(Square);
```


With new code it's very complex to override **toObject** method in class so override with prototype.

Above new code is correct way to implement?

How to assign default value of **customBorder** property 

```
let square = new Square({
      width: 100,
      height: 100,
      fill: "green",
      customBorder: true
    });

console.log(square.toObject());
```

Above **square.toObject()** return **customBorder** **false** even assigned from instance.

**Reply by ShaMan123** [reply]
Why not?
```ts
export class Square extends fabric.Rect {
  customBorder: boolean = false;
  constructor(options: Record<string, unknown>){
    super(options);
  }
 toObject(propsToInclude) {
  return super.toObject([["customBorder"], ...propertiesToInclude])
 }
}
```

**Reply by shaktippatel** [reply]
Getting below Error

```
Property 'toObject' in type 'Square' is not assignable to the same property in base type 'Rect<Partial<RectProps>, SerializedRectProps, ObjectEvents>'.
  Type '(propsToInclude: any) => Pick<Omit<Partial<RectProps> & TClassProperties<this>, keyof SerializedRectProps>, any> & SerializedRectProps' is not assignable to type '<T extends Omit<Partial<RectProps> & TClassProperties<this>, keyof SerializedRectProps>, K extends keyof T = never>(propertiesToInclude?: K[] | undefined) => Pick<...> & SerializedRectProps'.
    Type 'Pick<Omit<Partial<RectProps> & TClassProperties<this>, keyof SerializedRectProps>, any> & SerializedRectProps' is not assignable to type 'Pick<T, K> & SerializedRectProps'.
      Type 'Pick<Omit<Partial<RectProps> & TClassProperties<this>, keyof SerializedRectProps>, any> & SerializedRectProps' is not assignable to type 'Pick<T, K>'.ts(2416)
```

**Reply by shaktippatel** [reply]
> I don't see a problem When saying reproduction we mean this: https://codesandbox.io/p/sandbox/fabric-nextjs-sandbox-forked-z7l2nx?file=%2Fpages%2Findex.tsx&selection=%5B%7B%22endColumn%22%3A32%2C%22endLineNumber%22%3A25%2C%22startColumn%22%3A32%2C%22startLineNumber%22%3A25%7D%5D

getting below error in beta3 due to protected method
`Property 'toObject' is protected and only accessible within class 'FabricObject<Props, SProps, EventSpec>' and its subclasses.ts(2445)`

**Reply by shaktippatel** [reply]
What is expected date release of fabric v6?

**Reply by shaktippatel** [reply]
> Why not?
> 
> ```ts
> export class Square extends fabric.Rect {
>   customBorder: boolean = false;
>   constructor(options: Record<string, unknown>){
>     super(options);
>   }
>  toObject(propsToInclude) {
>   return super.toObject([["customBorder"], ...propertiesToInclude])
>  }
> }
> ```

I think the version used demo is old. after using latest version above method not working.

**Comment by shaktippatel** [comment]
Update fabricjs version in codesandbox demo and checked it not working with latest version.

https://codesandbox.io/p/sandbox/fabric-nextjs-sandbox-forked-d0ec8p?file=%2Fpackage.json&selection=%5B%7B%22endColumn%22%3A30%2C%22endLineNumber%22%3A13%2C%22startColumn%22%3A30%2C%22startLineNumber%22%3A13%7D%5D

**Reply by ShaMan123** [reply]
I don't have time to look at this this week
But these are simply ts errors. Who cares? We are still in beta.
Codesandbox has a problem installing deps. Not related to fabric

**Comment by gloriousjob** [comment]
The below is example of extending a rectangle by adding a label property (I could've called it LabeledRect but eh...). This is based on how the classes extend in the repo. However, I do get red at the top, like it can't reach into the classes to get the exports. If that could be resolved, this could probably be a starting point.
```ts
import { Rect, classRegistry } from "fabric";
import type { ObjectEvents } from "fabric/dist/src/EventTypeDefs";
import type { TProps } from "fabric/dist/src/shapes/Object/types";
import type { SerializedRectProps } from "fabric/dist/src/shapes/Rect";
import type { RectProps } from "fabric/dist/src/shapes/Rect";
import type { TClassProperties } from "fabric/dist/src/typedefs";

export const myRectDefaultValues: Partial<TClassProperties<MyRect>> = {
	label: 'Inside'
} as const;

interface UniqueMyRectProps {
	label: string;
}

export interface SerializedMyRectProps extends SerializedRectProps, UniqueMyRectProps {}

export interface MyRectProps extends RectProps, UniqueMyRectProps {}

const MY_RECT_PROPS = ['label'] as const;

export class MyRect<
		Props extends TProps<MyRectProps> = Partial<MyRectProps>,
		SProps extends SerializedMyRectProps = SerializedMyRectProps,
		EventSpec extends ObjectEvents = ObjectEvents
	>
	extends Rect<Props, SProps, EventSpec>
	implements MyRectProps
{
	declare label: string;

	static cacheProperties = [...Rect.cacheProperties, ...MY_RECT_PROPS];
	
	static ownDefaults: Record<string, any> = myRectDefaultValues;
	
	static getDefaults(): Record<string, any> {
	  return {
		...super.getDefaults(),
		...MyRect.ownDefaults,
	  };
	}

	constructor(options: Props) {
		super(options);
	}

	// not critical for classing but provides the code that the label will use
	_render(ctx: CanvasRenderingContext2D): void {
		super._render(ctx);

		if (this.label) {
			ctx.save();
	
			ctx.textAlign = 'center';
			ctx.textBaseline = 'middle';
			ctx.fillText(this.label, 0, 0);

			ctx.restore();
		}
	}

	toObject<
    T extends Omit<Props & TClassProperties<this>, keyof SProps>,
    K extends keyof T = never
  >(propertiesToInclude: K[] = []): Pick<T, K> & SProps {
    return { ...super.toObject(propertiesToInclude), ...MY_RECT_PROPS };
  }
}

classRegistry.setClass(MyRect);
classRegistry.setSVGClass(MyRect);
```

**Reply by ShaMan123** [reply]
> I don't have time to look at this this week But these are simply ts errors. Who cares? We are still in beta. Codesandbox has a problem installing deps. Not related to fabric

same applies.
The import paths are still wrong and need to be fixed
